docs/proposals/etcd-druid.md

```diff
@@ -0,0 +1,24 @@
+# Integrating etcd-druid with Gardener.
+
+Etcd is currently deployed by garden-controller-manager as a Statefulset. The sidecar container spec contains details pertaining to cloud-provider object-store which is injected into the statefulset via a mutable webhook running as part of the gardener extension [story](https://github.com/gardener/gardener/blob/master/docs/extensions/controlplane-webhooks.md#what-needs-to-be-implemented-to-support-a-new-cloud-provider).  This approach restricts the operations on etcd such as scale-up and upgrade. Etcd-druid will eliminate the need to hijack statefulset creation to add cloudprovider details. It has been designed to provide an intricate control over the procedure of deploying and maintaining etcd. The roadmap for etcd-druid can be found [here](https://github.com/gardener/etcd-druid/issues/2). etcd-
+
+This document explains how Gardener deploys etcd and what resources it creates for etcd-druid to deploy an etcd cluster.
+
+## Resources required by etcd-druid (created by Gardener)
+
+* Secret containing credentials to access backup bucket in Cloud provider object store.
+* TLS server and client secrets for etcd and backup-sidecar
+* Etcd CRD resource that contains parameters pertaining to etcd, backup-sidecar and cloud-provider object store.
+
+When an etcd resource is created in the cluster, the druid acts on it by creating an etcd statefulset, a service and a configmap containing etcd bootstrap script. The secrets containing the infrastructure credentials and the TLS certificates are mounted as volumes. 
+
+Garden-controller-manager, being cloud agnostic, deploys the etcd resource. This will not contain any cloud-specific information other than the cloud-provider. The extension controller that contains the cloud specific implementation to create the backup bucket will create it if needed and create a secret containing the credentials to access the bucket. The secret will have to be made available in the namespace the etcd statefulset will be deployed. If etcd and backup-sidecar communicates over TLS then the CA certificates, server and client certificates, and keys will also have to be made available in the namespace as well. The etcd resource will have reference to these aforementioned secrets. etcd-druid will deploy the statefulset only if the secrets are available.
+
+## Workflow
+* etcd-druid will be deployed and etcd CRD will be created as part of the seed bootstrap.
+* Garden-controller-manager creates backupBucket extension resource. Extension controller creates the backup bucket associated with the seed.
+* Garden-controller-manager creates backupentry associated with each shoot in the seed namespace. 
+* Garden-controller-manager creates   
+* etcd-druid acts on the etcd resource; druid creates the statefulset, the service and the configmap.
+
+![etcd-druid](./druid_integration.png)
```

Comment:
In the picture, what I don't understand as well is why the etcd-druid is registered as extension. IMHO that's wrong, and the etcd-druid should be deployed as part of the seed bootstrap or part of the shoot reconciliation flow. Will there be one or many etcd-druids in the seed?

Also, what about the reconciliation trigger? Let's assume we already have an etcd-druid and a bunch of etcd CRDs living. Now, in the next release of Gardener, we deploy a new etcd-druid. Will it immediately reconcile the CRDs, potentially rolling out a new etcd `StatefulSet`? If yes, we have to prevent this. We want Gardener to keep the control, and we want Gardener to trigger the reconciliation explicitly at a point in time it decides. For the extension controllers we have solved this with the operation annotation, see https://github.com/gardener/gardener/blob/master/docs/extensions/reconcile-trigger.md, and I think we need the same here as well.

Reply:
The idea was to have an etcd-druid per seed acting on etcd resources in the shoot namespaces. 
Related to shoots already deployed, etcd-druid should adopt the statefulset. I shall incorporate the annotation logic so that Gardener decides when the etcd statefulset gets adopted.

Reply:
Thanks! Please update the picture accordingly.